src/librustdoc/rustdoc-proc-macro/src/lib.rs

```diff
@@ -0,0 +1,57 @@
+use proc_macro::TokenStream;
+use quote::quote;
+use syn;
+
+#[proc_macro_derive(ItemTemplate)]
+pub fn item_template_derive(input: TokenStream) -> TokenStream {
+    let ast = syn::parse::<syn::DeriveInput>(input).unwrap();
+    let name = &ast.ident;
+    let impl_disp_capt_a_b_capt_cx = quote! {
+        impl std::fmt::Display + rustc_data_structures::captures::Captures<'a> + 'b + rustc_data_structures::captures::Captures<'cx>
+    };
+    let gen = quote! {
+        impl<'a, 'cx: 'a> ItemTemplate<'a, 'cx> for #name<'a, 'cx> {
+            fn item_and_mut_cx(&self) -> (&'a crate::clean::Item, std::cell::RefMut<'_, &'a mut crate::html::render::Context<'cx>>) {
+                (self.it, self.cx.borrow_mut())
+            }
+        }
+
+        impl<'a, 'cx: 'a> #name<'a, 'cx> {
+            fn document<'b>(&'b self) -> #impl_disp_capt_a_b_capt_cx {
+                display_fn(move |f| {
+                    let (item, mut cx) = self.item_and_mut_cx();
+                    let v = document(*cx, item, None, HeadingOffset::H2);
+                    write!(f, "{v}")
+                })
+            }
+
+            fn document_type_layout<'b>(&'b self) -> #impl_disp_capt_a_b_capt_cx {
```

Comment:
Will this method (and the two following it) always be used for all types (for example, for enums or functions)? If not we could consider adding parameters to `ItemTemplate`.

Reply:
> Will this method (and the two following it) always be used for all types (for example, for enums or functions)? If not we could consider adding parameters to `ItemTemplate`.

Not for all types. `document_type_layout()` is only used by typedef, enum, struct.

Reply:
Adding parameters might be a good idea here, will come back to this to see if it is worth the modification.